bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/NashornScriptEngineFactory.java

```diff
@@ -12,74 +12,67 @@
  */
 package org.openhab.core.automation.module.script.internal;
 
-import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Set;
 
 import javax.script.ScriptEngine;
-import javax.script.ScriptEngineManager;
 import javax.script.ScriptException;
 
 import org.openhab.core.automation.module.script.ScriptEngineFactory;
 import org.osgi.service.component.annotations.Component;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 /**
+ * An implementation of {@link ScriptEngineFactory} with customizations for Nashorn ScriptEngines.
  *
  * @author Simon Merschjohann - Initial contribution
+ * @author Scott Rushworth - removed default methods provided by ScriptEngineFactory
  */
 @Component(service = ScriptEngineFactory.class)
-public class NashornScriptEngineFactory implements ScriptEngineFactory {
-    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+public class NashornScriptEngineFactory extends AbstractScriptEngineFactory {
 
-    private ScriptEngineManager engineManager = new ScriptEngineManager();
+    private static final String SCRIPT_TYPE = "js";
 
     @Override
-    public List<String> getLanguages() {
-        return Arrays.asList("js", "javascript", "application/javascript");
+    public void logFactoryDetails() {
+        logger.info("Added customized ScriptEngineFactory for Nashorn");
     }
 
     @Override
-    public void scopeValues(ScriptEngine engine, Map<String, Object> scopeValues) {
-        Set<String> expressions = new HashSet<String>();
+    public List<String> getScriptTypes() {
+        List<String> scriptTypes = new ArrayList<>();
 
-        for (Entry<String, Object> entry : scopeValues.entrySet()) {
-            engine.put(entry.getKey(), entry.getValue());
+        for (javax.script.ScriptEngineFactory f : engineManager.getEngineFactories()) {
+            List<String> extensions = f.getExtensions();
 
-            if (entry.getValue() instanceof Class) {
-                expressions.add(String.format("%s = %s.static;", entry.getKey(), entry.getKey()));
+            if (extensions.contains(SCRIPT_TYPE)) {
+                scriptTypes.addAll(extensions);
+                scriptTypes.addAll(f.getMimeTypes());
             }
         }
-        String scriptToEval = String.join("\n", expressions);
-        try {
-            engine.eval(scriptToEval);
-        } catch (ScriptException e) {
-            logger.error("ScriptException while importing scope: {}", e.getMessage());
-        }
+        logger.trace("getScriptTypes(): {}", scriptTypes);
```

Comment:
I assume this logging can be removed.
If we want to be informed about the script types, why not moving this logging statement to the caller, so we get that information for all implementations?

Reply:
This logging was handy in testing/debugging, so I left it in. The method is called twice in ScriptEngineManagerImpl, to generate an interable when adding and removing SEFs (rare)...
```
for (String scriptType : engineFactory.getScriptTypes()) {
```
So, I'll need to set a variable and pull it out of there to do the logging. It won't take much, but is it worth it to move the logging to the caller?

Reply:
Sure, this one logging statement does not really harm.
But there will be a lot of them in the whole codebase if we keep all the "has been useful on testing" in the code.

The other point is:
There is an interface `ScriptEngineFactory`.
This factories are used by "someone".
An implementation needs to provide a `getScriptTypes()` method.
We provide an `AbstractScriptEngineFactory` that contians an implementation of that method and adds some logging.
If an implementation does not use the abstract base class there is no logging or a different one.
So, if we would like to be able to be informed about all the script types in the runtime, this works only for the subclasses of the abstract class but not for other implementations as long as the logging is in the abstract class.
If there is no need to get function calls traced anymore in production because it does not bring any value (as we know the method is working), IMHO we should remove it.
To use trace to detect some wrong behaviour on runtime systems of customers, I would like the same behaviour for all `ScriptEngineFactory` implementations (regardless if abstract base class is used) -- so my comment about moving it to the caller.

So, do you think we need to trace the if the abstract class is able to collect its types or which types are received by the caller.

Reply:
> So, if we would like to be able to be informed about all the script types in the runtime, this works only for the subclasses of the abstract class but not for other implementations as long as the logging is in the abstract class.

That makes perfect sense. I think this logging could help in troubeshooting issues, so I will move it to the caller.

Reply:
Hm. If you start openhab with logging level INFO you will see exactly nothing in the resulting 1MB log file. If openHAB does not yet have a way to tell about the status of its services, then we should tackle that (I suggested a service status like a thing status once). But adding log statements makes the situation worse. Just my 2ct.

Reply:
It is common practice to use debug and trace log to be able to diagnose some stuff at runtime.
This is not openHAB Core specific.
Some of that logging really makes sense.

Reply:
Most services are state machines though and it would be sufficient to be able to retrieve the current (error) state. In my vision of openHAB I would never need to go through logs, except to look up historic unexpected problems (errors), but not to find out why a particular thing does not work right now.

This is unrelated to the PR though.